test/topology/conftest.py

```diff
@@ -168,18 +171,43 @@ async def manager_internal(event_loop, request):
     await manager_int.stop()  # Stop client session and close driver after last test
 
 @pytest.fixture(scope="function")
-async def manager(request, manager_internal):
+async def manager(request, manager_internal, record_property, mode):
     """Per test fixture to notify Manager client object when tests begin so it can
     perform checks for cluster state.
     """
     test_case_name = request.node.name
+    run_id = request.config.getoption('run_id')
+    xml_path = pathlib.Path(request.config.getoption('xmlpath'))
+    test_log = xml_path.parent.parent / f"{xml_path.name[:-10]}.{test_case_name}.log"
+
+    await manager_internal.mark_scylla_log(test_case_name, test_log)
     await manager_internal.before_test(test_case_name)
     yield manager_internal
     # `request.node.stash` contains a flag stored in `pytest_runtest_makereport`
     # that indicates test failure.
     failed = request.node.stash[FAILED_KEY]
+    if failed:
+        # Save scylladb logs for failed tests in a separate directory and copy XML report to the same directory to have
+        # all related logs in one dir.
+        # Then add property to the XML report with the path to the directory, so it can be visible in Jenkins
+        failed_test_dir_path = xml_path.parent.parent / "failed_test" / f"{test_case_name}.{mode}.{run_id}"
+        await manager_internal.gather_related_logs(failed_test_dir_path, test_log)
+        logs_link = build_logs_dir_link(failed_test_dir_path)
+        record_property("SCYLLA_SERVER_LOGS", logs_link)
+    os.unlink(test_log)
     await manager_internal.after_test(test_case_name, not failed)
 
+
+def build_logs_dir_link(failed_test_dir_path: pathlib.Path):
+    """Build the link to the logs directory in artifacts for the failed test in a Jenkins report"""
```

Comment:
This code does not belong inside test.py.
And it doesn't apply when we run outside Jenkins.

test.py should not even know what "Jenkins" is

Reply:
How we can add Jenkins-specific features without knowing that we are on Jenkins? :thinking: 
This code will be executed regardless of where it was started: dev's machine or Jenkins. The only functionality that differs is URL generation which will be set to N/A in case when run was on the dev's machine.

Reply:
> How we can add Jenkins-specific features without knowing that we are on Jenkins? 🤔

The code that is specific to Jenkins pipeline implementation belongs inside the Jenkins pipeline implementation.

If the pipeline needs additional parameters to pass to test.py, you can add those parameters to test.py. But test.py doesn't know who is calling it and who is passing the parameters. It could be a developer running locally or it could be automation, Jenkins or something else.

Suppose that in the future we want to modify jenkins pipeline to change how the paths are named. If test.py contains code that depends on jenkins pipeline implementation details, such as path naming, the test.py code will break. On the other hand, if it provides configuration options that jenkins pipeline uses, then you will simply modify the values of those parameters in the same jenkins pipeline commit where you change path names.

Or, suppose that in the future we'd like to execute test.py from within GitHub actions instead of Jenkins. Instead of adding yet more code to test.py to support GitHub actions, you don't need to do anything, because test.py is already providing necessary configuration options, you can just provide appropriate values when calling it from GitHub actions.

It's a software engineering principle to write code in modularized and layered way. If layer X depends on layer Y, then layer Y does not know about the implementation of layer X. In fact layer X doesn't know about implementation of Y either, it only knows the interface of Y.